### PR TITLE
Add text and SVG mobjects to OpenGL

### PR DESCRIPTION
**Pull request by eulertour**
## Changelog / Overview

Added OpenGL-compatible text and SVG mobjects

## Motivation

Add text and svg mobjects that work with the OpenGL renderer

## Explanation for Changes

Created `opengl_text_mobject.py` and `opengl_svg_path.py` that are mostly copied from the Cairo versions. `opengl_svg_path.py` and `opengl_svg_mobject.py` have just enough information to allow reuse of our existing SVG / Text logic.

## Testing Status

None, but I'm willing to add some if I can figure out how to generate test data for OpenGL scenes.

## Further Comments

Regarding the amount of duplicated code that this and the earlier OpenGL changes introduced, I don't know if it'll ever make sense to combine our Cairo and OpenGL logic. If so, it will have to be done very carefully.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Review by naveen521kk — COMMENTED** [review]
This looks neat thanks. Tried it locally and it works. While I wonder something isn't working out
```py
class WriteStuff(Scene):
    def construct(self):
        example_text = OpenGLTex("This is a some text", tex_to_color_map={"text": YELLOW})
        example_tex = OpenGLMathTex(
            "\\sum_{k=1}^\\infty {1 \\over k^2} = {\\pi^2 \\over 6}",
        )
        group = OpenGLVGroup(example_text, example_tex)
        group.arrange(DOWN)
        group.width = config["frame_width"] - 2 * LARGE_BUFF

        self.play(Write(example_text))
        self.play(Write(example_tex))
        self.wait()
````
same as in `basic.py`. This creates a video like 

https://user-images.githubusercontent.com/49693820/113149602-d72cab80-9250-11eb-8d3f-2dfb7cd82cc7.mp4

while when using Cairo renderer from `basic.py` it gives


https://user-images.githubusercontent.com/49693820/113149672-e6135e00-9250-11eb-9b1c-1fe490de3e66.mp4

Is this difference indented?

File: manim/mobject/svg/opengl_text_mobject.py [hunk 1]
```diff
@@ -0,0 +1,1526 @@
...
+            mobject_without_dots[i].add(*[k for k in mobject[i] if k.__class__ != Dot])
+    else:
+        mobject_without_dots.add(*[k for k in mobject if k.__class__ != Dot])
+    if iscode:
+        code.code = mobject_without_dots
+        return code
+    return mobject_without_dots
+
+
+class OpenGLCairoText(OpenGLSVGMobject):
```
[the remark this answers is not in the document]
Comment (eulertour) [reply manim/mobject/svg/opengl_text_mobject.py:107]: I just ported everything from text_mobject.py.

[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/mobject/svg/opengl_text_mobject.py:107]: This is removed now, https://github.com/ManimCommunity/manim/commit/fdb7ecba9f16f84e669bcbd0fd9ee83228c32a92 you can remove `OpenGLCairoText` I think.

**Comment by eulertour** [comment]
The extra argument to `Mobject.rotate` was causing the characters to be displaced but I don't know why. Graphical unit testing is also difficult since we can't reproduce the output of the fonts on the GHA servers (this is consistent with cairo rendering).

**Comment by kilacoda** [comment]
`WriteStuff` works for me as well. However, when I tried to create a custom mobject subclassed from `OpenGLMathTex` (basically a [`ChemObject`](https://github.com/raghavg123/chanim/blob/master/chem_objects.py#L26) renamed to `OpenGLChemObject`), for one input string, I get a long traceback which ends up in `SVGMobject.handle_transforms`:

<details>
<summary>Traceback</summary>

```py
➜ manim render .\opengl_test.py -pqm --renderer=opengl
Manim Community v0.5.0
1: ChanimTest
2: WriteStuff

Choose number corresponding to desired scene/arguments.
(Use comma separated list for multiple entries)
Choice(s):  1
Traceback (most recent call last):
╭──────────────────────────────────────────────────────────────────────────────────────╮
│ File "C:\CodeProjects\Python\Manim\manim\manim\cli\render\commands.py", line 115, in │
│render                                                                                │
│    112             try:                                                              │
│    113                 renderer = OpenGLRenderer()                                   │
│    114                 scene = SceneClass(renderer)                                  │
│  ❱ 115                 scene.render()                                                │
│    116             except Exception:                                                 │
│    117                 console.print_exception()                                     │
│    118     elif config.renderer == "webgl":                                          │
│ File "C:\CodeProjects\Python\Manim\manim\manim\scene\scene.py", line 179, in render  │
│     176         """                                                                  │
│     177         self.setup()                                                         │
│     178         try:                                                                 │
│  ❱  179             self.construct()                                                 │
│     180         except EndSceneEarlyException:                                       │
│     181             pass                                                             │
│     182         self.tear_down()                                                     │
│ File "opengl_test.py", line 21, in construct                                         │
│    18                                                                                │
│    19 class ChanimTest(Scene):                                                       │
│    20     def construct(self):                                                       │
│  ❱ 21         c = OpenGLChemObject(Benzene_Diazonium_Chloride)                       │
│    22                                                                                │
│    23         self.play(Write(c))                                                    │
│    24         self.wait()                                                            │
│ File "c:\codeprojects\python\manim\chanim\chanim\chem_objects.py", line 143, in      │
│__init__                                                                              │
│    140             node_style=node_style,                                            │
│    141             bond_style=bond_style,                                            │
│    142         )                                                                     │
│  ❱ 143         super().__init__(                                                     │
│    144             "\\chemfig{%s}" % chem_code,                                      │
│    145             stroke_width=stroke_width,                                        │
│    146             tex_template=self.template,                                       │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\opengl_tex_mobject.py",   │
│line 400, in __init__                                                                 │
│    397         self.tex_environment = tex_environment                                │
│    398         tex_strings = self.break_up_tex_strings(tex_strings)                  │
│    399         self.tex_strings = tex_strings                                        │
│  ❱ 400         OpenGLSingleStringMathTex.__init__(                                   │
│    401             self,                                                             │
│    402             self.arg_separator.join(tex_strings),                             │
│    403             tex_environment=self.tex_environment,                             │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\opengl_tex_mobject.py",   │
│line 237, in __init__                                                                 │
│    234             environment=self.tex_environment,                                 │
│    235             tex_template=self.tex_template,                                   │
│    236         )                                                                     │
│  ❱ 237         OpenGLSVGMobject.__init__(                                            │
│    238             self,                                                             │
│    239             file_name=file_name,                                              │
│    240             should_center=should_center,                                      │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\opengl_svg_mobject.py",   │
│line 31, in __init__                                                                  │
│    28             "should_subdivide_sharp_curves": should_subdivide_sharp_curves,    │
│    29             "should_remove_null_curves": should_remove_null_curves,            │
│    30         }                                                                      │
│  ❱ 31         OpenGLVMobject.__init__(                                               │
│    32             self, stroke_width=stroke_width, fill_opacity=fill_opacity, **kwarg│
│    33         )                                                                      │
│    34         self.move_into_position(width, height)                                 │
│ File                                                                                 │
│"C:\CodeProjects\Python\Manim\manim\manim\mobject\types\opengl_vectorized_mobject.py",│
│line 118, in __init__                                                                 │
│     115                                                                              │
│     116         self.needs_new_triangulation = True                                  │
│     117         self.triangulation = np.zeros(0, dtype="i4")                         │
│  ❱  118         super().__init__(**kwargs)                                           │
│     119         self.refresh_unit_normal()                                           │
│     120                                                                              │
│     121     def get_group_class(self):                                               │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\opengl_mobject.py", line 91,  │
│in __init__                                                                           │
│      88         self.init_uniforms()                                                 │
│      89         self.init_updaters()                                                 │
│      90         # self.init_event_listners()                                         │
│  ❱   91         self.init_points()                                                   │
│      92         self.init_colors()                                                   │
│      93                                                                              │
│      94         self.shader_indices = None                                           │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\opengl_svg_mobject.py",   │
│line 37, in init_points                                                               │
│    34         self.move_into_position(width, height)                                 │
│    35                                                                                │
│    36     def init_points(self):                                                     │
│  ❱ 37         self.generate_points()                                                 │
│    38                                                                                │
│    39     def path_string_to_mobject(self, path_string: str, style: dict):           │
│    40         return OpenGLSVGPathMobject(                                           │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 128,│
│in generate_points                                                                    │
│    125         """                                                                   │
│    126         doc = minidom_parse(self.file_path)                                   │
│    127         for svg in doc.getElementsByTagName("svg"):                           │
│  ❱ 128             mobjects = self.get_mobjects_from(svg, {})                        │
│    129             if self.unpack_groups:                                            │
│    130                 self.add(*mobjects)                                           │
│    131             else:                                                             │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 174,│
│in get_mobjects_from                                                                  │
│    171             pass  # TODO, handle style                                        │
│    172         elif element.tagName in ["g", "svg", "symbol", "defs"]:               │
│    173             result += it.chain(                                               │
│  ❱ 174                 *[                                                            │
│    175                     self.get_mobjects_from(                                   │
│    176                         child, style, within_defs=within_defs or is_defs      │
│    177                     )                                                         │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 175,│
│in <listcomp>                                                                         │
│    172         elif element.tagName in ["g", "svg", "symbol", "defs"]:               │
│    173             result += it.chain(                                               │
│    174                 *[                                                            │
│  ❱ 175                     self.get_mobjects_from(                                   │
│    176                         child, style, within_defs=within_defs or is_defs      │
│    177                     )                                                         │
│    178                     for child in element.childNodes                           │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 174,│
│in get_mobjects_from                                                                  │
│    171             pass  # TODO, handle style                                        │
│    172         elif element.tagName in ["g", "svg", "symbol", "defs"]:               │
│    173             result += it.chain(                                               │
│  ❱ 174                 *[                                                            │
│    175                     self.get_mobjects_from(                                   │
│    176                         child, style, within_defs=within_defs or is_defs      │
│    177                     )                                                         │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 175,│
│in <listcomp>                                                                         │
│    172         elif element.tagName in ["g", "svg", "symbol", "defs"]:               │
│    173             result += it.chain(                                               │
│    174                 *[                                                            │
│  ❱ 175                     self.get_mobjects_from(                                   │
│    176                         child, style, within_defs=within_defs or is_defs      │
│    177                     )                                                         │
│    178                     for child in element.childNodes                           │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 203,│
│in get_mobjects_from                                                                  │
│    200                                                                               │
│    201         result = [m for m in result if m is not None]                         │
│    202         if config["renderer"] == "opengl":                                    │
│  ❱ 203             self.handle_transforms(element, OpenGLVGroup(*result))            │
│    204         else:                                                                 │
│    205             self.handle_transforms(element, VGroup(*result))                  │
│    206         if len(result) > 1 and not self.unpack_groups:                        │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\svg\svg_mobject.py", line 513,│
│in handle_transforms                                                                  │
│    510                 matrix[:, 1] *= -1                                            │
│    511                                                                               │
│    512                 for mob in mobject.family_members_with_points():              │
│  ❱ 513                     mob.points = np.dot(mob.points, matrix)                   │
│    514                 mobject.shift(x * RIGHT + y * UP)                             │
│    515                                                                               │
│    516             elif op_name == "scale":                                          │
│ File "C:\CodeProjects\Python\Manim\manim\manim\mobject\mobject.py", line 526, in     │
│__getattr__                                                                           │
│     523             return types.MethodType(setter, self)                            │
│     524                                                                              │
│     525         # Unhandled attribute, therefore error                               │
│  ❱  526         raise AttributeError(f"{type(self).__name__} object has no attribute │
│     527                                                                              │
│     528     @property                                                                │
│     529     def width(self):                                                         │
╰──────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: OpenGLTexSymbol object has no attribute 'points'
```

</details>

<details>
<summary>Code used</summary>

```py
class ChanimTest(Scene):
    def construct(self):
        Benzene_Diazonium_Chloride = "*6(-=-=(-\\chemabove{N_2}{\quad\scriptstyle+}\\chemabove{Cl}{\quad\scriptstyle-})-=-)"
        c = OpenGLChemObject(Benzene_Diazonium_Chloride)

        self.play(Write(c))
        self.wait()
```
</details>

while for another, the output is pretty jagged:
<details>
<summary> Code 2 </summary>

```py
class ChanimTest(Scene):
    def construct(self):
        Benzene = "*6(-=-=-=)"
        c = OpenGLChemObject(Benzene)

        self.play(Write(c))
        self.wait()
```
</details>

<details>
<summary> Output rendered with `-pqk`</summary>


https://user-images.githubusercontent.com/65204531/114291320-15716880-9aa4-11eb-95e3-0c1c74f0039a.mp4


</details>

I tested it out on some other input strings as well and got the same error. From what it seems, it breaks whenever it encounters an SVG with notmal text apparently. Is there a way around this?

**Comment by eulertour** [comment]
I added a missing check for the crash. The other problem might be solvable if you upload an SVG to compare against with cairo.

**Comment by kilacoda** [comment]
Great! The fix works.

Here are two videos comparing output from both renderers both rendered with `-k`, i.e. at 2160p60:
OpenGL:

https://user-images.githubusercontent.com/65204531/114293127-41471b00-9ab1-11eb-8070-3f2fc769c147.mp4

Cairo:

https://user-images.githubusercontent.com/65204531/114293129-42784800-9ab1-11eb-8239-125f2a999d0a.mp4

and here's the SVG being used:

[benzene_SVG.zip](https://github.com/ManimCommunity/manim/files/6291209/benzene_SVG.zip)

**Comment by eulertour** [comment]
I don't get the jagged video, what animation are you using? I generated this with Create.

https://user-images.githubusercontent.com/43117506/114293789-e9c39380-9a4d-11eb-8ea8-375e5b33ae37.mp4

**Comment by kilacoda** [comment]
I used `Write`.

**Comment by eulertour** [comment]
I get this with Write

https://user-images.githubusercontent.com/43117506/114293846-7c643280-9a4e-11eb-8ad4-d2aaf5abc65e.mp4

**Comment by kilacoda** [comment]
Weird... I can't get output like that from either `Write` or `Create`, even when explicitly using an `OpenGLSVGMobject` instead of the `OpenGLChemObject`. I still get the jagged, thin-lined version I sent earlier.

**Comment by Darylgolden** [comment]
It works correctly for me too:
https://user-images.githubusercontent.com/9102529/114294950-83ce1f00-9ad4-11eb-8611-44a96553d480.mp4

**Comment by Darylgolden** [comment]
@kilacoda Are you able to render the rest of the OpenGL example scenes correctly? Also, it seems that the Cairo renderer renders the SVG with a black stroke color while the OpenGL renderer renders it white, is this intended?

**Comment by kilacoda** [comment]
> @kilacoda Are you able to render the rest of the OpenGL example scenes correctly?

I tried them out just now, and apparently `InteractiveDevelopment` isn't working/not displaying anything in the preview window.

What code are you guys using exactly to render the SVG btw? I can try using that as is.

**Comment by Darylgolden** [comment]
```
from manim import *
from manim.opengl import *

# config["background_color"] = WHITE

class Test(Scene):
    def construct(self):
        self.play(
            Create(OpenGLSVGMobject("benzene.svg"))
        )
```
For Cairo I uncommented the background color line, since for some reason the colors are opposite on both renderers.

**Comment by eulertour** [comment]
It's pretty clear that the problem being discussed now isn't related to this PR, so can this be approved and the debugging moved into #1240?

**Review by kilacoda — APPROVED** [review]

**Comment by kilacoda** [comment]
There weren't any tests for this PR, is @ManimCommunity/tests okay with that?

**Comment by kilacoda** [comment]
Also, this seems to not have been squashed and merged...

**Pull request merged**